Remove Serialization namespace, and introduce relationship interface

drwill-ms opened the pull request.

Commit:
Remove Serialization namespace, and introduce relationship interface

drwill-ms commented on sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs:
```diff
@@ -830,12 +836,17 @@ public virtual async Task<Response<T>> GetRelationshipAsync<T>(string digitalTwi
+            string relationshipId,
+            GetRelationshipOptions options = null,
+            CancellationToken cancellationToken = default)
+            where T : IDigitalTwinRelationship
```
Introducing a constraint on type T

drwill-ms commented on sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs:
```diff
@@ -2024,32 +2042,5 @@ internal static string[] GetAuthorizationScopes(Uri endpoint)
 
             throw new InvalidOperationException($"Azure digital twins instance endpoint '{endpoint.AbsoluteUri}' is not valid.");
         }
-
```
Moved this helper to its own file.

drwill-ms commented on sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/BasicRelationship.cs:
```diff
@@ -46,41 +52,34 @@ namespace Azure.DigitalTwins.Core.Serialization
         /// </summary>
         [JsonExtensionData]
-        public IDictionary<string, object> CustomProperties { get; set; } = new Dictionary<string, object>();
+        public IDictionary<string, object> Properties { get; set; } = new Dictionary<string, object>();
```
Renamed to Properties

drwill-ms commented on sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/IDigitalTwinRelationship.cs:
```diff
@@ -0,0 +1,31 @@
+    /// <summary>
+    /// The service-defined fields of a digital twin relationship.
+    /// </summary>
+    public interface IDigitalTwinRelationship
```
Introducing an interface for relationship.

timtay-microsoft replied:
What are your thoughts on adding constants here that tell users what the name of each json property is? I'm considering doing so in Java

timtay-microsoft replied:
```csharp
const DigitalTwinRelationshipIdJsonPropertyName = "$relationshipId"
```

timtay-microsoft replied:
It's important that we convey these json property names somewhere, after all

drwill-ms replied:
I don't think a constant helps them any more than a comment, because they can't refer to the comment in an attribute, can they?

drwill-ms replied:
Also, with our current runtime, we can't put default impl on interfaces yet. That is too new of a .NET feature.

drwill-ms replied:
Found constants can be in an attribute. I've added that info in comments. Will leave it this way for GA, and we can always add constants later if they are requested.

drwill-ms commented on sdk/digitaltwins/Azure.DigitalTwins.Core/src/StreamHelper.cs:
```diff
@@ -0,0 +1,51 @@
+        /// <param name="objectSerializer">Object serializer used to serialize/deserialize an object.</param>
+        /// <param name="cancellationToken">Then cancellation token.</param>
+        /// <returns>A binary representation of the object written to a stream. Requires disposing.</returns>
+        internal static async Task<MemoryStream> WriteToStreamAsync<T>(T obj, ObjectSerializer objectSerializer, CancellationToken cancellationToken)
```
Split it up into async and sync, so we don't go async from a sync method.

timtay-microsoft commented on sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/IDigitalTwinRelationship.cs:
```diff
@@ -0,0 +1,31 @@
+    public interface IDigitalTwinRelationship
+    {
+        /// <summary>
+        /// The unique Id of the relationship. This field is present on every relationship.
```
You could add a note here saying what the json property name is that this property should correspond to ("$relationshipId" for this property)

timtay-microsoft reviewed (approved):
With or without the constants for json property names defined, this is good

barustum commented on sdk/digitaltwins/Azure.DigitalTwins.Core/samples/DigitalTwinsClientSample/CustomRelationship.cs:
```diff
@@ -2,15 +2,14 @@
-    internal class CustomRelationship
+    internal class CustomRelationship : IDigitalTwinRelationship
     {
         [JsonPropertyName("$relationshipId")]
```
why not use DigitalTwinsJsonPropertyNames.RelationshipId instead? (Same for the other properties below)

barustum commented on sdk/digitaltwins/Azure.DigitalTwins.Core/src/Models/DigitalTwinsJsonPropertyNames.cs:
```diff
@@ -0,0 +1,34 @@
+        public const string RelationshipTargetId = "targetId";
+
+        /// <summary>
+        /// The JSON property name for the relationship Id field.
```
relationship **name** field

barustum reviewed (approved):
Looks good. Left a couple of comments.

timtay-microsoft committed:
Fix Azure.Core paging interceptor logic (#16279)

If AsPages is called with a continuation token or pageSizeHint, it should be propagated into the actual .AsPages call
